src/fs_storage.c

```diff
@@ -0,0 +1,77 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2012-2018 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <sys/statvfs.h>
+
+#include "swupd.h"
+
+/* recursively calculate size from a complete list of bundle_result structs */
+long calculate_size(char *bname, struct list *bundle_info, bool installed)
```

Comment:
As far as I understood you have all information you need to calculate the bundle size in the manifest struct. I'd use this so you won't need to duplicate everything before calculating the size.